src/App.js

```diff
@@ -25,44 +24,97 @@ const generateSquares = () => {
   return squares;
 }
 
-const App = () => {
+const updateCurrentPlayer = (currentPlayer, setCurrentPlayer) => {
+  const updatePlayer = currentPlayer === PLAYER_1 ? PLAYER_2:PLAYER_1
+  setCurrentPlayer(updatePlayer)
+}
 
-  // This starts state off as a 2D array of JS objects with
-  // empty value and unique ids.
+const App = () => {
   const [squares, setSquares] = useState(generateSquares());
+  const [currentPlayer, setCurrentPlayer] = useState(PLAYER_1)
+  const [winner, setWinner ] = useState(null)
 
-  // Wave 2
-  // You will need to create a method to change the square 
-  //   When it is clicked on.
-  //   Then pass it into the squares as a callback
+  const checkForWinner = (squares) => {
+    let x = 0
+    let o = 0
+    let empty = false
+    const possibilities = [
+      [squares[0][0].value, squares[0][1].value, squares[0][2].value],
+      [squares[1][0].value, squares[1][1].value, squares[1][2].value],
+      [squares[2][0].value, squares[2][1].value, squares[2][2].value],
+      [squares[0][0].value, squares[1][0].value, squares[2][0].value],
+      [squares[0][1].value, squares[1][1].value, squares[2][1].value],
+      [squares[0][2].value, squares[1][2].value, squares[2][2].value],
+      [squares[0][0].value, squares[1][1].value, squares[2][2].value],
+      [squares[0][2].value, squares[1][1].value, squares[2][0].value]
+    ]
 
+    for(let i = 0; i < possibilities.length; i++){
+      if(possibilities[i][0] === '' || possibilities[i][1] === '' || possibilities[i][2] === ''){
+        empty = true
+      }
+      if(possibilities[i][0] === possibilities[i][1] && possibilities[i][1] === possibilities[i][2]){
+        if(possibilities[i][0] === 'x'){
+          x += 1
+        }
+        if( possibilities[i][0] === 'o'){
+          o += 1
+        }
+      }
+    }
 
-  const checkForWinner = () => {
-    // Complete in Wave 3
-    // You will need to:
-    // 1. Go accross each row to see if 
-    //    3 squares in the same row match
-    //    i.e. same value
-    // 2. Go down each column to see if
-    //    3 squares in each column match
-    // 3. Go across each diagonal to see if 
-    //    all three squares have the same value.
+    if(x === 0 && o === 0 && empty){
+      setWinner(null)
+    }
+    if(x === 0 && o === 0 && !empty)
+      setWinner('Tie')
+    if( x > o){
+      setWinner('x')
+    }
+    if(x < o){
+      setWinner('o')
+    }
+  }
 
+  const updateSquaresByID = (squareID) => {
+    for(let i = 0; i < squares.length; i++){
+      for(let j = 0; j < squares[i].length; j++){
+        if(squares[i][j].id === squareID && squares[i][j].value === ''){
+          squares[i][j].value = currentPlayer
+          updateCurrentPlayer(currentPlayer, setCurrentPlayer)
+          setSquares(squares)
```

Comment:
It turns out this call has no effect. You are processing the `squares` data directly, flipping the space corresponding to the clicked square to the current player. But since the `squares` reference is the same reference that React was already holding onto, this call to `setSquares` doesn't trigger a re-render. But we _do_ see a re-render, so what's going on?

Since the current player gets flipped, that triggers the re-render. The `squares` collection held by React _was_ updated, so when it renders the next view, we see the changes, but the `setSquares` had nothing to do with it.

Recall the general pattern for container objects (arrays, dictionaries, and general objects) in React involves making a copy of the existing values into a new value and updating the value held by the copy. This allows React to see the change, and also ensures that all the code that was rendered together has a consistent view of the data that was part of the render.